File: leetcode/278-first-bad-version.cpp
/*
You are a product manager and currently leading a team to develop a new product. Unfortunately, the latest version of your product fails the quality check. Since each version is developed based on the previous version, all the versions after a bad version are also bad.

Suppose you have n versions [1, 2, ..., n] and you want to find out the first bad one, which causes all the following ones to be bad.

You are given an API bool isBadVersion(version) which will return whether version is bad. Implement a function to find the first bad version. You should minimize the number of calls to the API.

Credits:
Special thanks to @jianchao.li.fighter for adding this problem and creating all test cases.
*/

// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int low = 1, high = n;
        while (low < high) {
            auto mid = (high - low) / 2 + low;
            auto result = isBadVersion(mid);
            if (!result) {
                low = mid + 1;
            } else if (mid != 1 && !isBadVersion(mid - 1)) {
                return mid;
            } else {
                high = mid - 1;
            }
        }
        return low;
    }
};

File: leetcode/278-first-bad-version.rb
# The is_bad_version API is already defined for you.
# @param {Integer} version
# @return {boolean} whether the version is bad
# def is_bad_version(version):

# @param {Integer} n
# @return {Integer}
def first_bad_version(n)
  b = 1
  e = n
  while b < e
    mid = (e - b) / 2 + b
    if is_bad_version mid
      e = mid
    else
      b = mid + 1
    end
  end
  b
end
